src/main/java/org/fcrepo/storage/ocfl/DefaultOcflObjectSession.java

```diff
@@ -0,0 +1,477 @@
+                                    final OcflRepository ocflRepo,
+                                    final String ocflObjectId,
+                                    final Path objectStaging,
+                                    final ObjectMapper objectMapper,
+                                    final Runnable deregisterHook) {
+        this.sessionId = sessionId;
+        this.ocflRepo = ocflRepo;
+        this.ocflObjectId = ocflObjectId;
+        this.objectStaging = objectStaging;
+        this.objectMapper = objectMapper;
+        this.deregisterHook = deregisterHook;
+
+        this.versionInfo = new VersionInfo();
+        this.deletePaths = new HashSet<>();
+        this.ocflOptions = new OcflOption[] {OcflOption.MOVE_SOURCE, OcflOption.OVERWRITE};
+
+        this.rootResourceId = loadRootResourceId();
+    }
+
+    @Override
+    public String sessionId() {
+        return sessionId;
+    }
+
+    @Override
+    public String ocflObjectId() {
+        return ocflObjectId;
+    }
+
+    @Override
+    public synchronized void writeResource(final String resourceId, final ResourceContent content) {
+        enforceOpen();
+
+        final var paths = resolvePersistencePaths(resourceId, content.getHeaders());
+
+        final var contentPath = encode(paths.getContentFilePath());
+        final var headerPath = encode(paths.getHeaderFilePath());
+
+        deletePaths.remove(contentPath);
+        deletePaths.remove(headerPath);
+
+        content.getHeaders().setContentPath(contentPath.path);
+
+        final var contentDst = createStagingPath(contentPath);
+        write(content.getContentStream(), contentDst);
+
+        final var headerDst = createStagingPath(headerPath);
+        writeHeaders(content.getHeaders(), headerDst);
+    }
+
+    @Override
+    public synchronized void deleteContentFile(final String resourceId, final ResourceHeaders headers) {
+        enforceOpen();
+
+        final var headerPath = encode(PersistencePaths.headerPath(rootResourceId(), resourceId));
+
+        if (newInSession(headerPath)) {
+            deleteResource(resourceId);
+        } else {
+            final var existingHeaders = readHeaders(resourceId);
+
+            deletePaths.add(encode(existingHeaders.getContentPath()));
+
+            final var headerDst = createStagingPath(headerPath);
+            writeHeaders(headers, headerDst);
+        }
+    }
+
+    @Override
+    public synchronized void deleteHeaderFile(final String resourceId) {
+        enforceOpen();
+
+        if (Objects.equals(rootResourceId(), resourceId)) {
+            deleteResource(resourceId);
+        } else {
+            deletePaths.add(encode(PersistencePaths.headerPath(rootResourceId(), resourceId)));
+        }
+    }
+
+    @Override
+    public synchronized void deleteResource(final String resourceId) {
+        enforceOpen();
+
+        if (Objects.equals(rootResourceId(), resourceId)) {
+            deleteObject = true;
```

Comment:
I can't answer this question definitively because it's up to the behavior we want to support. As written, an object can be deleted and recreated within the same session. We can disallow this if we want to, but there isn't a technical reason to do so.

Reply:
It seems that given an existing resource, if you delete it, restore it, and read it in the same session, the read operation will fail
```
@Test
    public void restoreAtomicRdfResource() {
        final var resourceId = "info:fedora/foo/bar";
        final var session = sessionFactory.newSession(resourceId);

        final var contentStr = "Test";
        final var content = atomicContainer(resourceId, "info:fedora/foo", contentStr);

        write(session, content);

        final var stagedContent = session.readContent(resourceId);

        assertResourceContent(contentStr, content, stagedContent);

        session.commit();

        final var committedContent = session.readContent(resourceId);

        assertResourceContent(contentStr, content, committedContent);

        // Start a second session to delete and then recreate the resource
        final var session2 = sessionFactory.newSession(resourceId);
        session2.deleteResource(resourceId);

        final var contentStr2 = "Test more";
        final var content2 = atomicContainer(resourceId, "info:fedora/foo", contentStr2);

        write(session2, content2);

        // Fails here for me
        final var stagedContent2 = session2.readContent(resourceId);
        assertResourceContent(contentStr2, content2, stagedContent2);

        session2.commit();

        final var committedContent2 = session2.readContent(resourceId);

        assertResourceContent(contentStr2, content2, committedContent2);
    }
```
But going back to my original question, I see that on commit (assuming it got past the read) in this scenario it first purges the existing OCFL object and then pushes any staged changes into a new object, so the handling generally of deleting objects seems good

Reply:
Good catch, that's a bug. Will fix.